ProcessLib/UncoupledProcessesTimeLoop.cpp

```diff
@@ -544,6 +542,112 @@ bool UncoupledProcessesTimeLoop::setCoupledSolutions()
     return true;
 }
 
+double UncoupledProcessesTimeLoop::computeTimeSteppping(
+    const double prev_dt, double& t, std::size_t& accepted_steps,
+    std::size_t& rejected_steps)
+{
+    bool all_process_steps_accepted = true;
+    // Get minimum time step size among step sizes of all processes.
+    double dt = std::numeric_limits<double>::max();
+    for (std::size_t i = 0; i < _per_process_data.size(); i++)
+    {
+        auto& ppd = *_per_process_data[i];
+        const auto& timestepper = ppd.timestepper;
+        if (t > timestepper->end())
+        {
+            // skip the process that already reaches the ending time.
+            ppd.skip_time_stepping = true;
+            continue;
+        }
+
+        auto& time_disc = ppd.time_disc;
+        auto const& x = *_process_solutions[i];
+
+        const double solution_error =
+            (t == timestepper->begin())
+                ? 0.
+                : time_disc->getRelativeError(
+                      x, timestepper->getSolutionNormType());
+        if (!timestepper->next(solution_error))
+        {
+            // Not all processes have accepted steps.
+            all_process_steps_accepted = false;
+        }
+
+        if (timestepper->getTimeStep().dt() >
+            std::numeric_limits<double>::epsilon())
+        {
+            if (timestepper->getTimeStep().dt() < dt)
+            {
+                dt = timestepper->getTimeStep().dt();
+            }
+        }
+        else
+        {
+            ppd.skip_time_stepping = true;
+        }
+    }
+
+    bool is_initial_step = false;
+    // Reset the time step with the minimum step size, dt
+    // Update the solution of the previous time step in time_disc.
+    for (std::size_t i = 0; i < _per_process_data.size(); i++)
+    {
+        const auto& ppd = *_per_process_data[i];
+        auto& timestepper = ppd.timestepper;
+        timestepper->resetCurrentTimeStep(dt);
+
+        if (ppd.skip_time_stepping)
+            continue;
+
+        if (t == timestepper->begin())
+        {
+            is_initial_step = true;
```

Comment:
In the current implementation, all processes must start at the same time. I will consider other possibility later.